GAE/src/org/akvo/flow/rest/FormAssemblyRestService.java

```
@@ -0,0 +1,440 @@
+    public static final String BUCKET = "s3bucket";
+
+    @PostMapping(consumes = "application/json")
+    @ResponseBody
+    public String publishForm(@RequestBody SurveyDto surveyDto) {
+        SurveyGroup survey = new SurveyGroupDAO().getByKey(surveyDto.getSurveyGroupId());
+        long formId = surveyDto.getKeyId();
+        Survey form = assembleForm(surveyDto);
+        FormUploadXml formUploadXml = assembleXmlForm(survey, form);
+
+        if (!formUploadXml.getXmlContent().isEmpty()) {
+            log.info("Uploading " + formId);
+            UploadStatusContainer uc = uploadFormXML(
+                    formUploadXml.getFormIdFilename(),
+                    formUploadXml.getFormIdVersionFilename(),
+                    formUploadXml.getXmlContent());
+            if (uc.getUploadedZip1() && uc.getUploadedZip2()) {
+                formUploadSuccess(survey, form);
+                log.info("Completed form assembly for " + formId);
+                return "OK";
+            } else {
+                log.severe("Failed to upload assembled form, id " + formId + "\n" + uc.getMessage());
+                return "Error";
+            }
+        }
+        return "Error";
+    }
+
+    @Nonnull
+    private FormUploadXml assembleXmlForm(SurveyGroup survey, Survey form) {
+        Properties props = System.getProperties();
+        String alias = props.getProperty("alias");
+        String xmlAppId = props.getProperty("xmlAppId");
+        String appStr = (xmlAppId != null && !xmlAppId.isEmpty()) ? xmlAppId : SystemProperty.applicationId.get();
+        XmlForm jacksonForm = new XmlForm(form, survey, appStr, alias);
+        try {
+            Long formId = form.getObjectId();
+            return new FormUploadXml(Long.toString(formId), //latest version in plain filename
+                    formId + "v" + form.getVersion(), //archive copy
+                    PublishedForm.generate(jacksonForm));
+        } catch (IOException e) {
```

I think rather than catching the exception, rather throw an exception at the method level and if an exception happens then the server will return a 500 to the client and we can deal with showing the user an error message there.

---

Actually I think it would be a better idea to return a Response object like here https://github.com/akvo/akvo-flow/blob/master/GAE/src/org/akvo/flow/rest/ImageUploadRestService.java#L185, what do you think? I can put a 500 code.

---

After giving it more thought I realized this is not how we are supposed to return errors and the ImageUploadRestService is doing it wrong. I have changed to using ResponseStatusException and also returning the SurveyDto when success.

---

> After giving it more thought I realized this is not how we are supposed to return errors and the ImageUploadRestService is doing it wrong. I have changed to using ResponseStatusException and also returning the SurveyDto when success.

I had also thought to make a mental note (better yet an issue) to also handle this properly in the `ImageUploadRestService`